internal/extractors/aez_reader.go

```diff
@@ -1,70 +1,54 @@
-	defer close(target)
-	for sourcePacket := range source {
-		if sourcePacket.Error != nil {
-			target <- sourcePacket
-			continue
-		}
-		reader := bytes.NewReader(sourcePacket.Buffer)
-		if sourcePacket.SourceHeader.PacketSequenceControl.GroupingFlags() == innosat.SPStandalone {
-			switch {
-			case sourcePacket.TMHeader.IsHousekeeping():
-				var sid aez.SID
-				binary.Read(reader, binary.BigEndian, &sid)
-				targetPacket, err := instrumentHK(sid, reader)
-				sourcePacket.Data = targetPacket
-				sourcePacket.SID = sid
-				sourcePacket.Error = err
-				if reader.Len() == 0 {
-					sourcePacket.Buffer = []byte{}
-				}
-				target <- sourcePacket
-			}
-		}
-	}
-}
-
-func instrumentHK(sid aez.SID, buf io.Reader) (common.Exportable, error) {
-	var dataPackage common.Exportable
-	var err error
-	switch sid {
-	case aez.SIDSTAT:
-		stat := aez.STAT{}
-		err = stat.Read(buf)
-		dataPackage = stat
-	case aez.SIDHTR:
-		htr := aez.HTR{}
-		err = htr.Read(buf)
-		dataPackage = htr
-	case aez.SIDPWR:
-		pwr := aez.PWR{}
-		err = pwr.Read(buf)
-		dataPackage = pwr
-	case aez.SIDCPRUA:
-		cpru := aez.CPRU{}
-		err = cpru.Read(buf)
-		dataPackage = cpru
-	case aez.SIDCPRUB:
-		cpru := aez.CPRU{}
-		err = cpru.Read(buf)
-		dataPackage = cpru
-	}
-	return dataPackage, err
-}
+package extractors
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/innosat-mats/rac-extract-payload/internal/aez"
+	"github.com/innosat-mats/rac-extract-payload/internal/common"
+)
+
+// DecodeAEZ parses AEZ packages
+func DecodeAEZ(target chan<- common.DataRecord, source <-chan common.DataRecord) {
+	defer close(target)
+	for sourcePacket := range source {
+		reader := bytes.NewReader(sourcePacket.Buffer)
+		switch {
+		case sourcePacket.TMHeader.IsHousekeeping():
+			var sid aez.SID
+			binary.Read(reader, binary.BigEndian, &sid)
+			sourcePacket.SID = sid
+			exportable, err := instrumentHK(sid, reader)
+			sourcePacket = addData(sourcePacket, reader, exportable, err)
+		case sourcePacket.TMHeader.IsTransparentData():
+			var rid aez.RID
+			binary.Read(reader, binary.BigEndian, &rid)
+			sourcePacket.RID = rid
+			exportable, err := instrumentTransparentData(rid, reader)
+			sourcePacket = addData(sourcePacket, reader, exportable, err)
+		default:
+			sourcePacket.Error = errors.New("the TMHeader isn't recognized as either housekeeping or tranparent data")
+		}
+		target <- sourcePacket
+	}
+}
+
+func addData(sourcePacket common.DataRecord, reader *bytes.Reader, exportable common.Exportable, exportableErr error) common.DataRecord {
+	buf := make([]byte, reader.Len())
+	_, bufErr := reader.Read(buf)
+	if bufErr != nil && bufErr != io.EOF {
+		if exportableErr == nil {
+			sourcePacket.Error = bufErr
+		} else {
+			sourcePacket.Error = fmt.Errorf("%v | %v", exportableErr, bufErr)
+		}
+	} else {
+		sourcePacket.Error = exportableErr
+	}
+	sourcePacket.Data = exportable
+	sourcePacket.Buffer = buf
+	return sourcePacket
+}

```

Comment:
The main reason for the function was to not duplicate the assignment of `Error` `Data` and `Buffer` in the already a bit complex function where `addData` is used.

I don't know what could really go wrong with the reader, but it felt problematic not to handle errors from functions that return them. If I understand it correctly, not receiving errors from functions that produce them causes panic / stop of execution and that seemed like making the program brittle.

I was also considering buffers instead of readers, because readers allows seeking which is contrary to how we want to use them. It would make the copy-pasta if removing the function much less annoying.